pkg/services/annotations/annotationsimpl/xorm_store.go

```diff
@@ -64,9 +64,63 @@ func (r *xormRepositoryImpl) Add(ctx context.Context, item *annotations.Item) er
 		if _, err := sess.Table("annotation").Insert(item); err != nil {
 			return err
 		}
+		return r.synchronizeTags(ctx, item)
+	})
+}
+
+// AddMany inserts large batches of annotations at once.
+// It does not return IDs associated with created annotations, and it does not support annotations with tags. If you need this functionality, use the single-item Add instead.
+// This is due to a limitation with some supported databases:
+// We cannot correlate the IDs of batch-inserted records without acquiring a full table lock in MySQL.
+// Annotations have no other uniquifier field, so we also cannot re-query for them after the fact.
+// So, callers can only reliably use this endpoint if they don't care about returned IDs.
+func (r *xormRepositoryImpl) AddMany(ctx context.Context, items []annotations.Item) error {
+	hasTags := make([]annotations.Item, 0)
+	hasNoTags := make([]annotations.Item, 0)
+
+	for i, item := range items {
+		tags := tag.ParseTagPairs(item.Tags)
+		item.Tags = tag.JoinTagPairs(tags)
+		item.Created = timeNow().UnixNano() / int64(time.Millisecond)
+		item.Updated = item.Created
+		if item.Epoch == 0 {
+			item.Epoch = item.Created
+		}
+		if err := r.validateItem(&items[i]); err != nil {
+			return err
+		}
 
+		if len(item.Tags) > 0 {
+			hasTags = append(hasTags, item)
+		} else {
+			hasNoTags = append(hasNoTags, item)
+		}
+	}
+
+	return r.db.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		// We can batch-insert every annotation with no tags. If an annotation has tags, we need the ID.
+		if _, err := sess.Table("annotation").InsertMulti(hasNoTags); err != nil {
```

Comment:
How does this work under the hood? Does xorm start a transaction or just send individual statements?

Reply:
I took a look inside InsertMulti. It batches all the items into one big statement. This is good, because it's implicitly a transaction without calling START ... COMMIT. But, without a size limit on the call to InsertMulti, you run the risk of running over the limits in various databases. For example, SQLite has a limit of 99 placeholders per statement. Make sure to test this with *very* large statements and run them against all 3 supported databases. I think we're likely to need a limiter.

Reply:
I created a utility to do this in a separate PR:
https://github.com/grafana/grafana/pull/57632

Once that's in, we ought to be covered here by simply calling it instead of InsertMulti directly.

Reply:
Done, we're now using the utility.